base/utf.jl

```diff
@@ -0,0 +1,880 @@
+    return encode_to_8(UInt16, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function convert(::Type{UTF8String}, str::Vector{UInt32})
+    len = sizeof(str)
+    # handle zero length string quickly
+    len == 0 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf32(str, len>>>2)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt32, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function convert(::Type{UTF8String},  src::UTF32String)
+    str = reinterpret(UInt32, src.data)
+    len = sizeof(str) >>> 2
+    # handle zero length string quickly
+    len <= 1 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf32(str, len-1)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt32, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function encode_to_8{T<:Union(UInt16, UInt32)}(::Type{T}, str, len)
+    buf = Vector{UInt8}(len)
+    out::UInt = 0
+    pos::UInt = 0
+    @inbounds while out < len
+        ch::UInt32 = str[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle 0x80-0x7ff
+        elseif ch < 0x800
+            @output_utf8_2!(buf, out, ch)
+        # Handle 0x10000-0x10ffff (if input is UInt32)
+        elseif T == UInt32 && ch > 0xffff
+            @output_utf8_4!(buf, out, ch)
+        # Handle surrogate pairs
+        elseif is_surrogate_char(ch)
+            ch = get_supplementary(ch, str[pos += 1])
+            @output_utf8_4!(buf, out, ch)
+        # Handle 0x800-0xd7ff, 0xe000-0xffff UCS-2 characters
+        else
+            @output_utf8_3!(buf, out, ch)
+        end
+    end
+    UTF8String(buf)
+end
+
+#=
+"""
+@brief      Converts a UTF-8 encoded string to UTF-32 encoding
+
+@param[in]  str::Vector{UInt8}
+
+@return     ::UTF32String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF32String}, src::UTF8String)
+    str = src.data
+    # handle zero length string quickly
+    if sizeof(str) == 0 ; buf = Vector{Char}(1) ; @inbounds buf[1] = Char(0) ; return UTF32String(buf) ; end
+    # Validate UTF-8 encoding, and get number of words to create
+    cnts::UTF_String_Counts = check_string_utf8(str)
+    len = cnts.cntT
+    # Optimize case where no characters > 0x7f
+    cnts.flags == 0 && return UTF32String(setindex!(copy!(Vector{Char}(len+1), str), 0, len+1))
+    # has multi-byte UTF-8 sequences
+    buf = Vector{Char}(len+1)
+    buf[len+1] = Char(0) # NULL termination
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch::UInt32 = str[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = Char(ch)
```

Comment:
OK, will do... I'd put those in because I'd been informed in no uncertain terms that `Char`s are not integer!  I just benchmarked it, and actually, it has an effect on performance (something that should be fixed).